app/httpparsers/VatSubmissionHttpParser.scala

```diff
@@ -17,34 +17,48 @@
 package httpparsers
 
 import play.api.http.Status.{BAD_REQUEST, CONFLICT, OK}
-import play.api.libs.json.JsSuccess
+import play.api.libs.json.{JsArray, JsSuccess}
 import uk.gov.hmrc.http.{HttpReads, HttpResponse}
 
+// scalastyle:off
 object VatSubmissionHttpParser {
 
-  val CodeKey = "code"
-  val InvalidPayloadKey = "INVALID_PAYLOAD"
-  val InvalidSessionIdKey = "INVALID_SESSIONID"
-  val InvalidCredentialIdKey = "INVALID_CREDENTIALID"
-  val FormBundleIdKey = "formBundle"
+  val codeKey = "code"
+  val failuresKey = "failures"
+  val invalidPayloadKey = "INVALID_PAYLOAD"
+  val invalidSessionIdKey = "INVALID_SESSIONID"
+  val invalidCredentialIdKey = "INVALID_CREDENTIALID"
+  val invalidCorrelationIdKey = "INVALID_CORRELATIONID"
+  val formBundleIdKey = "formBundle"
 
   type VatSubmissionResponse = Either[VatSubmissionFailure, VatSubmissionSuccess]
 
   implicit object VatSubmissionHttpReads extends HttpReads[VatSubmissionResponse] {
     override def read(method: String, url: String, response: HttpResponse): VatSubmissionResponse = {
       response.status match {
         case OK =>
-          (response.json \ FormBundleIdKey).validate[String].asEither
+          (response.json \ formBundleIdKey).validate[String].asEither
             .map(VatSubmissionSuccess)
             .left.map(_ => VatSubmissionFailure(OK, "VAT submission API - no form bundle ID in response"))
         case BAD_REQUEST =>
-          (response.json \ CodeKey).validate[String] match {
-            case JsSuccess(InvalidPayloadKey, _) =>
-              Left(VatSubmissionFailure(BAD_REQUEST, "VAT Submission API - invalid payload"))
-            case JsSuccess(InvalidSessionIdKey, _) =>
-              Left(VatSubmissionFailure(BAD_REQUEST, "VAT Submission API - invalid Session ID"))
-            case JsSuccess(InvalidCredentialIdKey, _) =>
-              Left(VatSubmissionFailure(BAD_REQUEST, "VAT Submission API - invalid Credential ID"))
+          (response.json \ failuresKey).validate[JsArray] match {
+            case JsSuccess(array, _) =>
```

Comment:
This is a perfect example of what we were discussing yesterday on the all day call, where something could be abstracted out, but still return flow to the calling method (as opposed to just handing off flow to the method).

 i.e. move the creation of the error array into a separate private `JsValue => Seq[String]` def, then you can call that on line 61 to create the log string, clearing up the intended logic of the read method and negating the need to turn off ScalaStyle